### ISSUE TITLE
Flask.teardown_request always passes an exception on the first request

### ISSUE THREAD

**Issue by obskyr** [pallets/flask#1694]
If a function is set with the `@app.teardown_request` decorator (assuming your Flask instance name is `app`), it *always* gets passed an exception for the very first request to the server. This exception is a `KeyError` from Werkzeug (even though that particular error is handled in Werkzeug's code).

Try copying and running the following bare-bones flask app:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from flask import Flask

import sys
import traceback

app = Flask(__name__)

@app.teardown_request
def teardown_request(exception):
    if exception:
        print "teardown_request received an exception:"
        traceback.print_exception(*sys.exc_info())
    else:
        print "Request went through without passing an exception."

@app.route("/")
def works_but_not_really():
    return "This gets returned, so this function finishes."

if __name__ == '__main__':
    app.run('localhost', 5000)
```

Running this and performing two GET requests in a row to it results in the following output:

```
 * Running on http://localhost:5000/ (Press CTRL+C to quit)
teardown_request received an exception:
Traceback (most recent call last):
  File "C:\Python27\lib\site-packages\werkzeug\serving.py", line 647, in inner
    fd = int(os.environ['WERKZEUG_SERVER_FD'])
  File "C:\Python27\lib\os.py", line 425, in __getitem__
    return self.data[key.upper()]
KeyError: 'WERKZEUG_SERVER_FD'
127.0.0.1 - - [14/Jan/2016 14:57:30] "GET / HTTP/1.1" 200 -
Request went through without passing an exception.
127.0.0.1 - - [14/Jan/2016 15:03:32] "GET / HTTP/1.1" 200 -
```

Not only GET requests cause this - at the very least a POST will do the same, and I assume all other types will too. This is a problem in case the teardown request performs a database rollback on receiving an exception - because it means any changes the first request may make to the database are silently (since the page is still returned just fine) ignored.

**Comment by ThiefMaster** [comment]
I think a better way to fix this would be calling `sys.exc_clear()` at some point to get rid of any unrelated errors. The fact that a caught exception is being passed to `teardown_request` even smells a bit like a bug.

**Comment by samuzzal** [comment]
Can anyone please confirm whether this issue is fixed? I could not reproduce this issue with Flask 0.11.1 and Python 2.7 on OS X and Fedora 23. I see that the PR https://github.com/pallets/werkzeug/pull/871 still open though.

#### Steps followed
* Run @obskyr 's sample code
```
[sam@localhost ~]$ chmod +x test.py 
[sam@localhost ~]$ ./test.py 
 * Running on http://localhost:5000/ (Press CTRL+C to quit)
```
* Perform two consecutive GET requests -
```
[sam@localhost ~]$ curl http://localhost:5000/; curl http://localhost:5000/
```
* Response received -
```
 * Running on http://localhost:5000/ (Press CTRL+C to quit)
Request went through without passing an exception.
127.0.0.1 - - [26/Sep/2016 14:22:30] "GET / HTTP/1.1" 200 -
Request went through without passing an exception.
127.0.0.1 - - [26/Sep/2016 14:22:30] "GET / HTTP/1.1" 200 -
```

**Comment by obskyr** [comment]
Okay, so here's the low-down. Despite being on Flask 0.10.1 and not having the error occur earlier (something weird was up there, I hope I misremembered something and it's not some sort of weird caching thing) I can now say this:

After uninstalling and reinstalling a lot of permutations of Flask and Werkzeug, this is my conclusion: the error **occurs on Flask 0.10.1 even with the newest version of Werkzeug**. However, **as of Flask 0.11.0** the error does not occur anymore, regardless of Werkzeug version. My tests haven't been super scientific, but this is the conclusion I've arrived at.

**Comment by samuzzal** [comment]
I got that exception now with Flask (0.10.1) and Werkzeug (0.11.11).

I am wondering whether or not to try out any fix for this. Especially when the issue is not reproducible with Flask 0.11.1 and it is only the first request that fails with 0.10.0.

**Comment by untitaker** [comment]
This is fixed in 0.11, most likely by #1393

**Comment by obskyr** [comment]
Ah, fantastic. Thanks for clearing that up. Incidentally that was one of the few things that stopped me from fully recommending Flask to people - after all, you shouldn't have to mess with weird workarounds to do basic stuff. All good now!
